## Week 8 Instructor-led Lab: Advanced Data Manipulation  
By: Dean Wilson  
Course: BGEN 632: Graduate Introduction to Python  
Date: 20250421  

In [2]:
# import modules
import os
import numpy as np
import pandas as pd

In [3]:
# set working directory
os.chdir('C:\\Users\\deanb\\OneDrive\\Desktop\\Python Graduate Class\\Week 8\\week8labs-main\\data')  # add your filepath within the quotes
os.getcwd()  # confirm change

'C:\\Users\\deanb\\OneDrive\\Desktop\\Python Graduate Class\\Week 8\\week8labs-main\\data'

In [4]:
gh_turnover = pd.read_csv("gh_turnover_four_projects.csv") # Open the file
gh_turnover.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   project_id     473 non-null    int64  
 1   domain         473 non-null    object 
 2   language       473 non-null    object 
 3   windows        473 non-null    int64  
 4   window_idx     473 non-null    int64  
 5   num_team       473 non-null    int64  
 6   num_commits    473 non-null    int64  
 7   blau_gender    473 non-null    float64
 8   Gini_gh_ten    473 non-null    float64
 9   user_id        473 non-null    int64  
 10  Core1          473 non-null    int64  
 11  commits        473 non-null    int64  
 12  propCommits    473 non-null    float64
 13  gender         449 non-null    object 
 14  github_tenure  473 non-null    int64  
 15  leavesNextQ    473 non-null    int64  
dtypes: float64(3), int64(10), object(3)
memory usage: 59.3+ KB


### Renaming Columns

In [34]:
gh_turnover = gh_turnover.rename(columns = 
                                 {
                                     'domain':'project_domain',
                                     'language': 'project_language',
                                     'windows': 'project_age',
                                     'window_idx': 'quarter',
                                     'num_team': 'team_size',
                                     'num_commits': 'project_commits',
                                     'blau_gender': 'gender_blau',
                                     'Gini_gh_ten': 'gh_tenure_gini',
                                     'Core1': 'core_dev',
                                     'commits': 'user_commits',
                                     'propCommits': 'commit_proportion',
                                     'github_tenure': 'gh_tenure',
                                     'leavesNextQ': 'leaves_next_q'
                                 }
                                ) # Selected Columns Renamed
gh_turnover.info() # Verifying Renaming

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   project_id         473 non-null    int64  
 1   project_domain     473 non-null    object 
 2   project_language   473 non-null    object 
 3   project_age        473 non-null    int64  
 4   quarter            473 non-null    int64  
 5   team_size          473 non-null    int64  
 6   project_commits    473 non-null    int64  
 7   gender_blau        473 non-null    float64
 8   gh_tenure_gini     473 non-null    float64
 9   user_id            473 non-null    int64  
 10  core_dev           473 non-null    int64  
 11  user_commits       473 non-null    int64  
 12  commit_proportion  473 non-null    float64
 13  gender             449 non-null    object 
 14  gh_tenure          473 non-null    int64  
 15  leaves_next_q      473 non-null    int64  
dtypes: float64(3), int64(10), 

### Data Wrangling

In [35]:
gh_turnover['user_commits'].mean() # Calculating Mean of user_commits

np.float64(339.36575052854124)

In [36]:
gh_turnover['user_commits'].median() # Calculating Median of user_commits

np.float64(18.0)

In [37]:
turnover_new = gh_turnover.filter(regex='(^p)|(g)') # Filtering for columns starting with p or containing g
turnover_new.info() # Verify filter

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   project_id        473 non-null    int64  
 1   project_domain    473 non-null    object 
 2   project_language  473 non-null    object 
 3   project_age       473 non-null    int64  
 4   project_commits   473 non-null    int64  
 5   gender_blau       473 non-null    float64
 6   gh_tenure_gini    473 non-null    float64
 7   gender            449 non-null    object 
 8   gh_tenure         473 non-null    int64  
dtypes: float64(2), int64(4), object(3)
memory usage: 33.4+ KB


In [38]:
turnover_new.query('project_commits > 9000') # Selecting rows where project_commits is more than 9,000

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

In [39]:
turnover_newer = ( 
    gh_turnover 
    .filter(regex='(^p)|(g)')
    .query('project_commits > 9000')
) # Chaining the filter and query in turnover_newer df
turnover_newer # Verify changes

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

In [40]:
turnover_df1 = (
    gh_turnover
    .filter(regex='(e$)|(s)')
    .query('gh_tenure >= 1991')
) # Created new DF from gh_turnover that filters for columns ending in e and containing s, as well as selecting rows with gh_tenure greater than or equal to 1,991
turnover_df1 # Verify changes

project_language  project_age  team_size  project_commits  user_id  \
339             Java           16          2                7    33272   
457           Python           23         20             3070    88459   

     user_commits  gh_tenure  leaves_next_q  
339             6       2006              0  
457             1       2054              0

In [41]:
print(turnover_df1['user_commits'].mean()) # Print filtered user_commits mean
print(turnover_df1['user_commits'].median()) # Print filtered user_commits median

3.5
3.5


### Results
* The mean and median of turnover_df1 are both 3.5 and differ significantly from the earlier mean of 339.37 and median of 18. This is due to turnover_df1 being filtered to rows where gh_tenure is greater than or equal to the value of 1,991. This resulted in only two rows where user commits were 6 and 1. The sample size has drastically dropped, explaining the massive difference between the old and new mean and median.

### References
* ChatGPT. (2025, April 21). Descriptive statistics analysis in Python [Large language model response]. ChatGPT. https://chatgpt.com/share/68068aae-6ffc-8001-9788-c45c4afcf5ed
* Ojeda, S. (2023, June 29). Calculating mean, median, and mode in Python. Stack Abuse. https://stackabuse.com/calculating-mean-median-and-mode-in-python/
* 08_Tutorial
* 08_Instructor_Led_Lab